File: packages/v2ray_box/linux/native_messaging.cc
#include "native_messaging.h"

#include "desktop_core.h"
#include "native_messaging_config.h"

#include <sys/stat.h>
#include <unistd.h>

#include <cerrno>
#include <vector>
#include <chrono>
#include <cstring>
#include <fstream>
#include <map>
#include <sstream>

namespace v2ray_box {
namespace {

constexpr int kExtensionPingMaxAgeSeconds = 120;

bool SetPrivateFileMode(const std::string& path, mode_t mode) {
  return chmod(path.c_str(), mode) == 0;
}

bool WritePrivateFile(const std::string& path, const std::string& content) {
  std::ofstream out(path, std::ios::trunc);
  if (!out.is_open()) {
    return false;
  }
  out << content;
  out.close();
  return SetPrivateFileMode(path, 0600);
}

bool FileExists(const std::string& path) {
  struct stat st {};
  return stat(path.c_str(), &st) == 0;
}

bool IsExecutable(const std::string& path) {
  return access(path.c_str(), X_OK) == 0;
}

std::string JsonEscape(const std::string& value) {
  std::string escaped;
  escaped.reserve(value.size() + 8);
  for (const char ch : value) {
    switch (ch) {
      case '\\':
        escaped += "\\\\";
        break;
      case '"':
        escaped += "\\\"";
        break;
      case '\n':
        escaped += "\\n";
        break;
      case '\r':
        escaped += "\\r";
        break;
      case '\t':
        escaped += "\\t";
        break;
      default:
        escaped += ch;
        break;
    }
  }
  return escaped;
}

std::string ChromeManifestPath() {
  const char* home = std::getenv("HOME");
  if (home == nullptr) {
    return "";
  }
  return std::string(home) +
         "/.config/google-chrome/NativeMessagingHosts/" +
         kNativeMessagingHostName + ".json";
}

std::string ChromiumManifestPath() {
  const char* home = std::getenv("HOME");
  if (home == nullptr) {
    return "";
  }
  return std::string(home) +
         "/.config/chromium/NativeMessagingHosts/" +
         kNativeMessagingHostName + ".json";
}

std::string FirefoxManifestPath() {
  const char* home = std::getenv("HOME");
  if (home == nullptr) {
    return "";
  }
  return std::string(home) + "/.mozilla/native-messaging-hosts/" +
         kNativeMessagingHostName + ".json";
}

bool EnsureParentDirectory(const std::string& path) {
  const auto pos = path.find_last_of('/');
  if (pos == std::string::npos) {
    return false;
  }
  return EnsureDirectory(path.substr(0, pos));
}

bool WriteNativeManifest(const std::string& path,
                         const std::string& host_path,
                         bool firefox) {
  if (!EnsureParentDirectory(path)) {
    return false;
  }

  std::ostringstream json;
  json << "{\n";
  json << "  \"name\": \"" << kNativeMessagingHostName << "\",\n";
  json << "  \"description\": \"Secure VPN proxy auth bridge\",\n";
  json << "  \"path\": \"" << JsonEscape(host_path) << "\",\n";
  json << "  \"type\": \"stdio\"\n";
  if (firefox) {
    json << "  ,\"allowed_extensions\": [ \"" << kFirefoxExtensionId << "\" ]\n";
  } else {
    json << "  ,\"allowed_origins\": [ \"chrome-extension://" << kChromeExtensionId
         << "/\" ]\n";
  }
  json << "}\n";
  return WritePrivateFile(path, json.str());
}

bool CopyBinary(const std::string& source, const std::string& destination) {
  std::ifstream in(source, std::ios::binary);
  if (!in.is_open()) {
    return false;
  }
  std::ofstream out(destination, std::ios::binary | std::ios::trunc);
  if (!out.is_open()) {
    return false;
  }
  out << in.rdbuf();
  out.close();
  return SetPrivateFileMode(destination, 0700);
}

std::string FindBundledNativeHostBinary() {
  const std::string exe_dir = GetExecutableDirectory();
  const std::vector<std::string> candidates = {
      JoinPath(exe_dir, "secure_vpn_native_host"),
      JoinPath(exe_dir, "../lib/secure_vpn_native_host"),
      JoinPath(exe_dir, "lib/secure_vpn_native_host"),
  };
  for (const auto& candidate : candidates) {
    if (IsExecutable(candidate)) {
      return candidate;
    }
  }
  return "";
}

int64_t ReadPingTimestamp() {
  std::ifstream in(NativeMessaging::ExtensionPingPath());
  if (!in.is_open()) {
    return 0;
  }
  int64_t value = 0;
  in >> value;
  return value;
}

}  // namespace

std::string NativeMessaging::NativeHostDirectory() {
  return JoinPath(GetWorkingDirectory(), "native_host");
}

std::string NativeMessaging::NativeHostBinaryPath() {
  return JoinPath(NativeHostDirectory(), "secure_vpn_native_host");
}

std::string NativeMessaging::SessionCredentialsPath() {
  return JoinPath(NativeHostDirectory(), "session.json");
}

std::string NativeMessaging::ExtensionPingPath() {
  return JoinPath(NativeHostDirectory(), "extension_ping");
}

bool NativeMessaging::InstallHost(const std::string& source_binary_path) {
  const std::string source =
      source_binary_path.empty() ? FindBundledNativeHostBinary()
                                 : source_binary_path;
  if (source.empty()) {
    return false;
  }
  if (!EnsureDirectory(NativeHostDirectory())) {
    return false;
  }
  return CopyBinary(source, NativeHostBinaryPath());
}

bool NativeMessaging::InstallManifests() {
  const std::string host_path = NativeHostBinaryPath();
  if (!IsExecutable(host_path)) {
    return false;
  }

  bool ok = true;
  const std::string chrome_path = ChromeManifestPath();
  const std::string chromium_path = ChromiumManifestPath();
  const std::string firefox_path = FirefoxManifestPath();
  if (!chrome_path.empty()) {
    ok = WriteNativeManifest(chrome_path, host_path, false) && ok;
  }
  if (!chromium_path.empty()) {
    ok = WriteNativeManifest(chromium_path, host_path, false) && ok;
  }
  if (!firefox_path.empty()) {
    ok = WriteNativeManifest(firefox_path, host_path, true) && ok;
  }
  return ok;
}

bool NativeMessaging::PublishCredentials(const std::string& host,
                                         int port,
                                         const std::string& username,
                                         const std::string& password) {
  if (!EnsureDirectory(NativeHostDirectory())) {
    return false;
  }

  std::ostringstream json;
  json << "{\n";
  json << "  \"type\": \"credentials\",\n";
  json << "  \"host\": \"" << JsonEscape(host) << "\",\n";
  json << "  \"port\": " << port << ",\n";
  json << "  \"username\": \"" << JsonEscape(username) << "\",\n";
  json << "  \"password\": \"" << JsonEscape(password) << "\"\n";
  json << "}\n";
  return WritePrivateFile(SessionCredentialsPath(), json.str());
}

bool NativeMessaging::ClearCredentials() {
  RemoveFileIfExists(SessionCredentialsPath());
  RemoveFileIfExists(ExtensionPingPath());
  return true;
}

std::map<std::string, bool> NativeMessaging::GetBrowserHelperStatus() {
  std::map<std::string, bool> status;
  const std::string host_path = NativeHostBinaryPath();
  status["hostInstalled"] = IsExecutable(host_path);

  const bool chrome_manifest = FileExists(ChromeManifestPath());
  const bool chromium_manifest = FileExists(ChromiumManifestPath());
  const bool firefox_manifest = FileExists(FirefoxManifestPath());
  status["manifestInstalled"] =
      chrome_manifest || chromium_manifest || firefox_manifest;
  status["chromeManifestInstalled"] = chrome_manifest;
  status["chromiumManifestInstalled"] = chromium_manifest;
  status["firefoxManifestInstalled"] = firefox_manifest;
  status["credentialsActive"] = FileExists(SessionCredentialsPath());

  const int64_t ping_ts = ReadPingTimestamp();
  const auto now = std::chrono::system_clock::now().time_since_epoch();
  const int64_t now_sec =
      std::chrono::duration_cast<std::chrono::seconds>(now).count();
  status["extensionConnected"] =
      ping_ts > 0 && (now_sec - ping_ts) <= kExtensionPingMaxAgeSeconds;

  status["ready"] = status["hostInstalled"] && status["manifestInstalled"] &&
                    status["extensionConnected"] && status["credentialsActive"];
  return status;
}

}  // namespace v2ray_box

File: packages/v2ray_box/macos/Classes/NativeMessaging.swift
import Foundation

enum NativeMessaging {
    private static let fileManager = FileManager.default

    static var nativeHostDirectory: URL {
        baseDirectory.appendingPathComponent("native_host", isDirectory: true)
    }

    static var nativeHostBinaryPath: URL {
        nativeHostDirectory.appendingPathComponent("secure_vpn_native_host", isDirectory: false)
    }

    static var sessionCredentialsPath: URL {
        nativeHostDirectory.appendingPathComponent("session.json", isDirectory: false)
    }

    static var extensionPingPath: URL {
        nativeHostDirectory.appendingPathComponent("extension_ping", isDirectory: false)
    }

    private static var baseDirectory: URL {
        let appSupport = fileManager.urls(for: .applicationSupportDirectory, in: .userDomainMask).first!
        return appSupport.appendingPathComponent("V2rayBox/working", isDirectory: true)
    }

    @discardableResult
    static func installHost(sourceBinaryPath: String? = nil) -> Bool {
        let source = sourceBinaryPath ?? findBundledNativeHostBinary()
        guard let source, !source.isEmpty else {
            return false
        }
        guard ensureDirectory(nativeHostDirectory) else {
            return false
        }
        let destination = nativeHostBinaryPath.path
        do {
            if fileManager.fileExists(atPath: destination) {
                try fileManager.removeItem(atPath: destination)
            }
            try fileManager.copyItem(atPath: source, toPath: destination)
            try fileManager.setAttributes([.posixPermissions: 0o700], ofItemAtPath: destination)
            return isExecutable(destination)
        } catch {
            return false
        }
    }

    @discardableResult
    static func installManifests() -> Bool {
        let hostPath = nativeHostBinaryPath.path
        guard isExecutable(hostPath) else {
            return false
        }

        var ok = true
        for (path, firefox) in manifestTargets() {
            ok = writeNativeManifest(at: path, hostPath: hostPath, firefox: firefox) && ok
        }
        return ok
    }

    @discardableResult
    static func publishCredentials(host: String, port: Int, username: String, password: String) -> Bool {
        guard ensureDirectory(nativeHostDirectory) else {
            return false
        }
        let json = """
        {
          "type": "credentials",
          "host": "\(jsonEscape(host))",
          "port": \(port),
          "username": "\(jsonEscape(username))",
          "password": "\(jsonEscape(password))"
        }
        """
        return writePrivateFile(path: sessionCredentialsPath.path, content: json + "\n")
    }

    @discardableResult
    static func clearCredentials() -> Bool {
        removeIfExists(sessionCredentialsPath.path)
        removeIfExists(extensionPingPath.path)
        return true
    }

    static func getBrowserHelperStatus() -> [String: Bool] {
        let hostInstalled = isExecutable(nativeHostBinaryPath.path)
        let chromeManifest = fileExists(chromeManifestPath())
        let chromiumManifest = fileExists(chromiumManifestPath())
        let edgeManifest = fileExists(edgeManifestPath())
        let firefoxManifest = fileExists(firefoxManifestPath())
        let manifestInstalled = chromeManifest || chromiumManifest || edgeManifest || firefoxManifest
        let credentialsActive = fileExists(sessionCredentialsPath.path)

        let pingTimestamp = readPingTimestamp()
        let now = Int64(Date().timeIntervalSince1970)
        let extensionConnected = pingTimestamp > 0 &&
            (now - pingTimestamp) <= Int64(NativeMessagingConfig.extensionPingMaxAgeSeconds)

        let ready = hostInstalled && manifestInstalled && extensionConnected && credentialsActive
        return [
            "hostInstalled": hostInstalled,
            "manifestInstalled": manifestInstalled,
            "chromeManifestInstalled": chromeManifest,
            "chromiumManifestInstalled": chromiumManifest || edgeManifest,
            "firefoxManifestInstalled": firefoxManifest,
            "credentialsActive": credentialsActive,
            "extensionConnected": extensionConnected,
            "ready": ready,
        ]
    }

    private static func homeDirectory() -> String {
        NSHomeDirectory()
    }

    private static func chromeManifestPath() -> String {
        "\(homeDirectory())/Library/Application Support/Google/Chrome/NativeMessagingHosts/\(NativeMessagingConfig.hostName).json"
    }

    private static func chromiumManifestPath() -> String {
        "\(homeDirectory())/Library/Application Support/Chromium/NativeMessagingHosts/\(NativeMessagingConfig.hostName).json"
    }

    private static func edgeManifestPath() -> String {
        "\(homeDirectory())/Library/Application Support/Microsoft Edge/NativeMessagingHosts/\(NativeMessagingConfig.hostName).json"
    }

    private static func firefoxManifestPath() -> String {
        "\(homeDirectory())/Library/Application Support/Mozilla/NativeMessagingHosts/\(NativeMessagingConfig.hostName).json"
    }

    private static func manifestTargets() -> [(String, Bool)] {
        [
            (chromeManifestPath(), false),
            (chromiumManifestPath(), false),
            (edgeManifestPath(), false),
            (firefoxManifestPath(), true),
        ]
    }

    private static func findBundledNativeHostBinary() -> String? {
        let bundle = Bundle(for: V2rayBoxPlugin.self)
        let candidates = [
            bundle.path(forResource: "secure_vpn_native_host", ofType: nil),
            Bundle.main.path(forResource: "secure_vpn_native_host", ofType: nil),
            Bundle.main.bundlePath + "/Contents/Resources/secure_vpn_native_host",
            Bundle.main.bundlePath + "/Contents/MacOS/secure_vpn_native_host",
            Bundle.main.bundlePath + "/Contents/Frameworks/secure_vpn_native_host",
        ]
        for candidate in candidates {
            if let path = candidate, isExecutable(path) {
                return path
            }
        }
        return nil
    }

    private static func writeNativeManifest(at path: String, hostPath: String, firefox: Bool) -> Bool {
        guard ensureParentDirectory(for: path) else {
            return false
        }
        let allowed = firefox
            ? "\"allowed_extensions\": [ \"\(NativeMessagingConfig.firefoxExtensionId)\" ]"
            : "\"allowed_origins\": [ \"chrome-extension://\(NativeMessagingConfig.chromeExtensionId)/\" ]"
        let json = """
        {
          "name": "\(NativeMessagingConfig.hostName)",
          "description": "RioNexTunnel proxy auth bridge",
          "path": "\(jsonEscape(hostPath))",
          "type": "stdio",
          \(allowed)
        }
        """
        return writePrivateFile(path: path, content: json + "\n")
    }

    private static func readPingTimestamp() -> Int64 {
        guard let content = try? String(contentsOf: extensionPingPath, encoding: .utf8) else {
            return 0
        }
        return Int64(content.trimmingCharacters(in: .whitespacesAndNewlines)) ?? 0
    }

    private static func ensureDirectory(_ url: URL) -> Bool {
        do {
            try fileManager.createDirectory(at: url, withIntermediateDirectories: true)
            return true
        } catch {
            return false
        }
    }

    private static func ensureParentDirectory(for path: String) -> Bool {
        let url = URL(fileURLWithPath: path).deletingLastPathComponent()
        return ensureDirectory(url)
    }

    private static func writePrivateFile(path: String, content: String) -> Bool {
        do {
            try content.write(toFile: path, atomically: true, encoding: .utf8)
            try fileManager.setAttributes([.posixPermissions: 0o600], ofItemAtPath: path)
            return true
        } catch {
            return false
        }
    }

    private static func removeIfExists(_ path: String) {
        if fileManager.fileExists(atPath: path) {
            try? fileManager.removeItem(atPath: path)
        }
    }

    private static func fileExists(_ path: String) -> Bool {
        var isDirectory: ObjCBool = false
        return fileManager.fileExists(atPath: path, isDirectory: &isDirectory) && !isDirectory.boolValue
    }

    private static func isExecutable(_ path: String) -> Bool {
        fileManager.isExecutableFile(atPath: path)
    }

    private static func jsonEscape(_ value: String) -> String {
        var escaped = ""
        for char in value {
            switch char {
            case "\\": escaped += "\\\\"
            case "\"": escaped += "\\\""
            case "\n": escaped += "\\n"
            case "\r": escaped += "\\r"
            case "\t": escaped += "\\t"
            default: escaped.append(char)
            }
        }
        return escaped
    }
}
